src/deferred.js

```diff
@@ -13,6 +13,23 @@ function Thrower( ex ) {
 	throw ex;
 }
 
+function adoptValue( value, resolve, reject ) {
+	var method;
+
+	// Check for promise aspect first to privilege synchronous behavior
+	if ( value && jQuery.isFunction( ( method = value.promise ) ) ) {
```

Comment:
And to clarify: `if ( jQuery.isFunction( value.then ) ) { value.then.call(…); }` is prohibited because it accesses `value.then` twice, and the results of those accesses won't be identical in the presence of a pathological getter like https://github.com/promises-aplus/promises-tests/blob/0d409df278345753c9527539a5558db67d42bb02/lib/tests/helpers/thenables.js#L29 .